### ISSUE TITLE
ccl/backupccl: contents of table after restore may be different than original contents when backup was taken

### ISSUE THREAD

**Issue by renatolabs** [cockroachdb/cockroach#101118]
I'm seeing some strange behavior of cluster backups taken _in mixed-version_ (there's a good chance "mixed-version" is relevant to the issue as I've only been seeing this for specific node versions). 

Here's the unexpected sequence of events:

**1. Take a cluster backup:**

> ```BACKUP INTO 'gs://cockroachdb-backup-testing/10_22.2.7-to-current_cluster_all-planned-on-current-executed-on-22.2.7_fAyS?AUTH=implicit' AS OF SYSTEM TIME '1681064488807127644.0000000000' WITH detached, revision_history, encryption_passphrase = '...'```

An incremental backup is taken as well shortly after (which may or may not be relevant as far as this issue is concerned, I can check whether I still get a failure with just a full backup with no incrementals).

> ```BACKUP INTO LATEST IN 'gs://cockroachdb-backup-testing/10_22.2.7-to-current_cluster_all-planned-on-current-executed-on-22.2.7_fAyS?AUTH=implicit' AS OF SYSTEM TIME '1681065145646539228.0000000000' WITH detached, revision_history, encryption_passphrase = '...'```

**2. Fingerprint one of the tables in that database at the time the backup was taken:**

>```SELECT * FROM [SHOW EXPERIMENTAL_FINGEPRINTS FROM TABLE tpcc.stock] AS OF SYSTEM TIME '1681065145646539228.0000000000'```

Output:

```
   index_name   |     fingerprint
----------------+----------------------
  stock_pkey    | 5318086895587186715
```

**3. Restore the backup taken above in a 23.1 cluster (release-23.1 and master have the same behavior):**

> ```RESTORE FROM LATEST IN 'gs://cockroachdb-backup-testing/10_22.2.7-to-current_cluster_all-planned-on-current-executed-on-22.2.7_fAyS?AUTH=implicit' WITH encryption_passphrase = '...'```

Restore finishes with no errors.

**4. Compute fingerprint of the same table:**

>```SELECT * FROM [SHOW EXPERIMENTAL_FINGEPRINTS FROM TABLE tpcc.stock]```

Output:

```
   index_name   |     fingerprint
----------------+----------------------
  stock_pkey    | 8078242823056593497
```

As you can see, **the fingerprints are different** and they should have been the same.

*****

#### Is it possible that this is an issue with `EXPERIMENTAL_FINGEPRINTS`?

Not in this case. I exported the data for that table that the backup should have backed up (using the same timestamp) and after restore; the fingerprints match the values obtained above. In addition, _the data is indeed different_. Here's an example where I loaded two tables with the contents of the exported data:

```sql
SELECT
    original.s_quantity AS original_quantity,
    restored.s_quantity AS restored_quantity
FROM
    original
INNER JOIN
    restored USING (s_i_id, s_w_id)
WHERE
    original.s_w_id = 98 AND original.s_i_id = 81185;
```

Note that `(s_w_id, s_i_id)` is tpcc's stock primary key [1].

The result of the query above shows a difference between the table at the time of the backup, and the restored table:

```
  original_quantity | restored_quantity
--------------------+--------------------
                100 |                10
```

#### Reproduction

This issue was found while running the `backup/mixed-version` roachtest as implemented in #100755. In particular, this failure is observed in somewhere between 10% to 30% of runs with seed `COCKROACH_RANDOM_SEED=4454232140074523958`. As the name in step 1 suggests, the backup is taken in mixed-version state: the backup is sent to a 23.1 node, but adopted by a 22.2.7 node.

Let me know if you need help reproducing the issue.

I also uploaded some artifacts from a failure I observed recently in a folder in the `cockroach-tmp` bucket [2]. Here's a breakdown of the files in that folder:

* `backup_22.2.7-to-current_cluster_all-planned-on-current-executed-on-22.2.7/`: the backup itself.
* `original_export`: exported contents of `tpcc.stock` using the same AOST timestamp as the one passed to the backup command. Importing this data will result in the same fingerprint as the one in step 2.
* `restored_export`: exported contents of `tpcc.stock` after the backup is restored. Importing this data will result in the same fingerprint as the one in step 4.
* `roachtest_artifacts`: the roachtest artifacts for a test failure, including cockroach logs.

**Note that the `cockroach-tmp` bucket has a 20-day TTL policy.**

[1] https://github.com/cockroachdb/cockroach/blob/cd20e09e3cbf17caecb565297178c4f0e10db945/pkg/workload/tpcc/ddls.go#L156

[2] <https://console.cloud.google.com/storage/browser/cockroach-tmp/backup_mixed_version_fingerprint_mismatch?project=cockroach-shared&supportedpurview=project&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false>

Jira issue: CRDB-26813

**Comment by msbutler** [comment]
Sounds good, one more q: have you ever seen this on DATABASE or TABLE backups? Or just cluster?

**Comment by renatolabs** [comment]
Good point -- this seems to only happen in cluster backups.

**Comment by msbutler** [comment]
The output of SHOW BACKUP on the backup _looks_ reasonable, and `SHOW BACKUP with check_files` passes. So according to the manifest, nothing seems off.... 

```
demo@127.0.0.1:26257/movr> SELECT * FROM [SHOW BACKUP FROM LATEST IN 'gs://cockroach-tmp/backup_mixed_version_fingerprint_mismatch/backup_22.2.7-to-current_cluster_all-planned-on-current-executed-on-22.2.7?AUTH=implicit' WITH
                        -> encryption_passphrase ='7u9FPtQJCnema7E1cwDPT0DjeOSpK3l6FOgtEZCfEOl2QTJjq7'] WHERE "database_name"='tpcc';
  database_name | parent_schema_name | object_name | object_type | backup_type |         start_time         |          end_time          | size_bytes |   rows   | is_full_cluster | regions
----------------+--------------------+-------------+-------------+-------------+----------------------------+----------------------------+------------+----------+-----------------+----------
  tpcc          | NULL               | public      | schema      | full        | NULL                       | 2023-04-09 18:21:28.807127 |       NULL |     NULL |        t        | NULL
  tpcc          | public             | warehouse   | table       | full        | NULL                       | 2023-04-09 18:21:28.807127 |    3309511 |      100 |        t        | NULL
  tpcc          | public             | district    | table       | full        | NULL                       | 2023-04-09 18:21:28.807127 |   12702477 |     1001 |        t        | NULL
  tpcc          | public             | customer    | table       | full        | NULL                       | 2023-04-09 18:21:28.807127 | 1918110029 |  3000059 |        t        | NULL
  tpcc          | public             | history     | table       | full        | NULL                       | 2023-04-09 18:21:28.807127 |  233209629 |  3061689 |        t        | NULL
  tpcc          | public             | order       | table       | full        | NULL                       | 2023-04-09 18:21:28.807127 |  170087548 |  3061677 |        t        | NULL
  tpcc          | public             | new_order   | table       | full        | NULL                       | 2023-04-09 18:21:28.807127 |   14015898 |   961678 |        t        | NULL
  tpcc          | public             | item        | table       | full        | NULL                       | 2023-04-09 18:21:28.807127 |    8251679 |   100000 |        t        | NULL
  tpcc          | public             | stock       | table       | full        | NULL                       | 2023-04-09 18:21:28.807127 | 3415892405 | 10000062 |        t        | NULL
  tpcc          | public             | order_line  | table       | full        | NULL                       | 2023-04-09 18:21:28.807127 | 1842046153 | 30614638 |        t        | NULL
  tpcc          | NULL               | public      | schema      | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |       NULL |     NULL |        t        | NULL
  tpcc          | public             | warehouse   | table       | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |     736756 |      100 |        t        | NULL
  tpcc          | public             | district    | table       | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |    2811100 |     1000 |        t        | NULL
  tpcc          | public             | customer    | table       | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |   15894943 |    27414 |        t        | NULL
  tpcc          | public             | history     | table       | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |     927535 |    13755 |        t        | NULL
  tpcc          | public             | order       | table       | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |    1479982 |    27489 |        t        | NULL
  tpcc          | public             | new_order   | table       | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |     315146 |    27489 |        t        | NULL
  tpcc          | public             | item        | table       | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |          0 |        0 |        t        | NULL
  tpcc          | public             | stock       | table       | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |   43706783 |   123526 |        t        | NULL
  tpcc          | public             | order_line  | table       | incremental | 2023-04-09 18:21:28.807127 | 2023-04-09 18:32:25.646539 |   16366794 |   275761 |        t        | NULL
(20 rows)
```

**Comment by renatolabs** [comment]
Running the test without any incremental backups revealed that:

* incrementals are unrelated, as the same issue still happened.
* turns out this is not related to cluster backups. I observed the same failure in a table backup.

The only thing in common across all failures (so far) is that this happens when the backup is planned on 23.1 and adopted by 22.2.7 (could still be a red herring, but worth highlighting).

**Comment by msbutler** [comment]
thanks Renato, that's really helpful to know. I think the leading thread to pull on is a potential incompatibility between a 22.2 export request caller (i.e. a backup sql processor)  and a 23.1 export request callee (a kv server). We've run into a couple data corruption failures over the cycle here. 

Going to chat with my team about this.

**Comment by renatolabs** [comment]
Ran the test with the same seed as the one reported in the original failure, but this time capturing a `debug.zip` after each backup is taken. The test failed to validate the same backup (and saw a fingerprint mismatch in the same table, tpcc.stock), so there's good reason to believe this is a very similar run hitting the same issue.

See `debug_10_22.2.7-to-current_cluster_all-planned-on-current-executed-on-22.2.7.zip` in the GCS folder linked in the issue description.

**Comment by renatolabs** [comment]
Updates: we have an ongoing [thread](https://cockroachlabs.slack.com/archives/C03JCUUSCD6/p1681323362933489) (internal), and some temporary changes in #101572 to help debug this issue.

I uploaded some more files to a subfolder in the GCS folder in the issue description (`annotated_failure`).

Description of all files in [annotated_failure](https://console.cloud.google.com/storage/browser/cockroach-tmp/backup_mixed_version_fingerprint_mismatch/annotated_failure?project=cockroach-shared&supportedpurview=project&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false):

* `1_22.2.8-to-current_cluster_all-planned-on-current-executed-on-22.2.8_h4Cq`: the backup itself.
* `after_restore_export`: export of the `bank` table after the backup was restored
* `bank_export`: export of the bank table right after the backup was taken (using same AOST timestamp)
* `debug_1_22.2.8-to-current_cluster_all-planned-on-current-executed-on-22.2.8.zip`: debug zip collected after backup was taken (including cockroach logs)
* `roachtest_artifacts`: the roachtest artifacts for the test failure

The test failed with the following error message:

```
mixed_version_backup.go:512: mismatched fingerprints for table bank.bank
--- Expected:
bank_pkey:-3950960582922845670
--- Actual:
bank_pkey:1481010650334147930
```

I once again restored both the sets of exported data (`bank_export` in an `original` table, and `after_restore_export` in a `restored` table) and compared them.

```
root@localhost:26257/bank> show experimental_fingerprints from table original;                                                                                                                         
   index_name   |     fingerprint
----------------+-----------------------
  original_pkey | -3950960582922845670

root@localhost:26257/bank> show experimental_fingerprints from table restored;                                                                                                                         
   index_name   |     fingerprint
----------------+----------------------
  restored_pkey | 1481010650334147930
```

Notice how the fingerprints match the values obtained during the test, so the exported data corresponds to the data the test saw when it failed.

```
root@localhost:26257/bank> select count(*) from original;                                                                                                                                              
  count
---------
  98642

root@localhost:26257/bank> select count(*) from restored;                                                                                                                                              
  count
---------
  98642
```

Both tables have the same number of rows. We can check which rows are different with the following query:

```
root@localhost:26257/bank> select original.id from original inner join restored using (id) 
where original.balance <> restored.balance or original.payload <> restored.payload;                         
  id
-------
  417
  960
```

It seems like, this time, two rows are different. We can further learn that their payloads are the same, but the associated balance is different:

```
root@localhost:26257/bank> select original.id, original.balance as original_balance, 
restored.balance as restored_balance, (original.payload = restored.payload) as payloads_match 
from original inner join restored using (id) where original.id IN (417, 960);         
  id  | original_balance | restored_balance | payloads_match
------+------------------+------------------+-----------------
  417 |           -13780 |            -4890 |       t
  960 |            59857 |             1948 |       t                                                                                                          
```

**Comment by msbutler** [comment]
I think I can tell coherent story of how this could happen, thanks to all the help from Renato, David, Steven, Aditya, and Rui: essentially, the new 23.1 generative split and scatter processor can't restore  paginated 22.2 backups with revision history that specifically produce an SST whose _last_ key isn't the latest MVCC value. I'm still trying to understand how 22.2 restore reacts to this.

*On the backup side*
On master and all previous release branches, in the file sink, we consider flushing the sst to disk every time we add a new export request to the sink. _If we extend_  the already buffered sst with a new export request, we will only flush the extended SST to[ if it's last  key is at a boundary](https://github.com/dt/cockroach/blob/7c8a3720171cd2a07280cd8b381c60f57b9ec798/pkg/ccl/backupccl/file_sst_sink.go#L341-#L344). 

But, _if we cannot extend_ the buffered sst because the incoming ssts' start key is less than the buffered sst's endkey, [we flush it, no questions asked](https://github.com/dt/cockroach/blob/7c8a3720171cd2a07280cd8b381c60f57b9ec798/pkg/ccl/backupccl/file_sst_sink.go#L287), before dealing with the incoming export request response. 

Because 22.2 used a shared file sink with the sorted queue logic, we could end up with the following buffered export responses in the top half of the queue, _post sort_: 

......{some more ssts sorted before b_start},[b_start,b_mid]

the bottom half of the queue would have [c_start,c_end].... and so on.

Now, when w[e drain](https://github.com/cockroachdb/cockroach/blob/7c8a3720171cd2a07280cd8b381c60f57b9ec798/pkg/ccl/backupccl/file_sst_sink.go#L185) the first half the queue, the last response will not get flushed since it is split mid key,  _but_  [we do remove it ](https://github.com/cockroachdb/cockroach/blob/7c8a3720171cd2a07280cd8b381c60f57b9ec798/pkg/ccl/backupccl/file_sst_sink.go#L200)from the queue! It just sits in an open file.

When it's time to drain the queue again, suppose [a_start,a_end] is at the top of the queue as it was added after the last drain, we will flush [b_start,b_mid], and b_mid will be the last key in the sst. 


**on the restore side**
coming soon...

**Comment by renatolabs** [comment]
This was fixed by #102179. Ran a batch of tests to verify its correctness, and no issues were found. Closing.
